text/projects/Lukasz_Selwa/Galil_Seiferas.tex

```diff
@@ -0,0 +1,308 @@
+Pokażemy za to sam algorytm wyznaczania $\HRP1(x)$ i $\HRP2(x)$ w odpowiednim czasie.
+
+\subsection{Wyznaczanie $k$-HRP}
+
+Przedstawimy jak wyznaczyć wszystkie $\HRP$ słowa $w$ o długości $n$.
+Załóżmy, że mamy policzony ciąg pierwszych $\HRP$ $v_1, v_2, ..., v_r$ i ich odpowiednie zakresy $[L_1, R_1], [L_2, R_2], ..., [L_r, R_r]$.
+W tedy z lematu \ref{Lem_1} wiemy, że kolejny $\HRP$ może pojawić się od pozycji $R_r / 2$. 
+
+Ustalmy pozycję $pos > R_r / 2$.
+Załóżmy, że sprawdziliśmy wszystkie pozycje w przedziale $[R_r / 2, pos)$ i wiemy, że żadna z nich nie jest $\HRP$.
+Niech $j$ będzie liczbą taką, że $w[1...j]$ to najdłuższy wspólny prefiks $w$ oraz $w[pos...n]$. 
+Jeśli $pos \cdot k \leq pos + j$ to $v[1...pos]$ jest $k$-HRP słowa $w$, ponieważ $(w[1...pos])^k$ jest prefiksem $w$ oraz $w$ jest słowem pierwotnym bo inaczej znaleźlibyśmy $k$-HRP wcześniej.
+Mamy też od razu obliczony zakres $v_{r+1}$, ponieważ z definicji $L_{r+1} = 2 \cdot pos$ oraz $R_{r+1} = pos + j$.
+Jeśli natomiast $pos \cdot k > pos + j$ to $w[1...pos]$ nie może być $k$-HRP.
+
+Zastanówmy się gdzie może być następny potencjalny koniec $k$-HRP.
+Jeśli $j$ nie należy do żadnego zakresu $[L_i, R_i]$ to z lematu \ref{Lem_2} najkrótszy okres $w[1...j]$ jest większy od $j/k$. W takim przypadku możemy więc przesunąć pozycję $pos$ o $\lfloor j / k \rfloor + 1$ i mieć pewność, że nie pominęliśmy kolejnego $k$-HRP.
+
+Jeśli natomiast $j \in [L_i, R_i]$ dla pewnego $k$-HRP $v_i$ to z lematu \ref{Lem_2} wiemy, że $p(w[1...j]) = L_i / 2$. Stąd następna potencjalna pozycja jest równa $pos + L_i / 2$. Ale ponieważ $j \geq L_i$ to wiemy, że najdłuższy prefiks $w$ i $w[pos...n]$ ma co najmniej długość $j - L_i / 2$. 
+Taka analiza prowadzi do następującego algorytmu.
+
+\begin{algorithm}[H]
+\caption{PREPROCESS}
+\label{preprocess}
+\begin{algorithmic} 
+\STATE $\textit{spcopes} \gets \texttt{empty list}$
+\STATE $pos \gets 1, j \gets 0$
+\WHILE{$pos + m < n$}
+    \WHILE{$pos + j < m \land w[j + 1] = w[pos + j + 1] $}
+        \STATE $j \gets j + 1$
+    \ENDWHILE
+    \IF{$k \cdot pos \leq pos + j$}
+        \STATE Append $[2 \cdot pos, pos + j]$ to list \textit{spcopes}.
+    \ENDIF
+    \IF{$j \in [L_i, R_i]$ for some $[L_i, R_i]$ in \textit{scopes}}
+        \STATE $pos \gets pos + L_i / 2$
+        \STATE $j \gets j - L_i / 2$
+    \ELSE
+        \STATE $pos \gets pos + \lfloor j / k \rfloor + 1$
+        \STATE $j \gets 0$
+    \ENDIF
+\ENDWHILE
+\RETURN \textit{scopes}
+\end{algorithmic}
+\end{algorithm}
+
+Podobnie jak w przypadku algorytmu \ref{simple-text-search} podany algorytm jest liniowy ponieważ cały czas rośnie wartość $k \cdot pos + j$.
+
+Zauważmy, że jeśli zatrzymamy algorytm gdy znajdziemy $\HRP1$ i $\HRP2$ to dostaniemy algorytm obliczający pierwsze dwa $\HRP$ w stałej pamięci i liniowy od długości $\HRP2$.
```

Comment:
Do I understand right that HRP1 is the first one appended to the list "scopes" and HRP2 is the second one?